File: dynamiclayout.ml
(* DynamicLayout OCaml API — generate layout JSON from OCaml code.
 *
 * Single-file module. Zero dependencies outside stdlib.
 * Works on OCaml 4.14+.
 *
 * Example:
 *   open DynamicLayout
 *   let b = begin_ "My Page" in
 *   let b = fieldset b "Info" in
 *   let b = label b "Hello from OCaml!" in
 *   let b = end_fieldset b in
 *   let b = button b "ok" "OK" "primary" true in
 *   print_endline (end_layout b)
 *)

type builder = {
  buf: Buffer.t;
  depth: int ref;
  comma: bool ref;
  cnt: int ref;
}

let putc b c = Buffer.add_char b.buf c; b
let puts b s = Buffer.add_string b.buf s; b
let indent b =
  let _ = putc b '\n' in
  for _ = 1 to !(b.depth) do ignore (puts b "  ") done; b

let cm b = if !(b.comma) then (putc b ','; b.comma := false; b) else b

let esc b s =
  let _ = putc b '"' in
  String.iter (function
    | '"' -> puts b "\\\""; | '\\' -> puts b "\\\\"; | '\n' -> puts b "\\n"
    | '\r' -> puts b "\\r"; | '\t' -> puts b "\\t"; | c -> putc b c) s;
  putc b '"'

let kv b k v =
  let _ = cm b in let _ = putc b ',' in let _ = indent b in
  let _ = esc b k in let _ = puts b ": " in let _ = esc b v in
  b.comma := true; b

let kvint b k v = kv b k (string_of_int v)
let kvbool b k = kv b k "true"

let obj b =
  let _ = cm b in let _ = putc b ',' in let _ = indent b in
  let _ = putc b '{' in b.depth := !(b.depth) + 1; b.comma := false; b

let close_obj b =
  b.depth := !(b.depth) - 1; let _ = indent b in let _ = putc b '}' in
  b.comma := true; b

let arr b k =
  let _ = cm b in let _ = putc b ',' in let _ = indent b in
  let _ = esc b k in let _ = puts b ": [" in
  b.depth := !(b.depth) + 1; b.comma := false; b

let close_arr b =
  b.depth := !(b.depth) - 1; let _ = indent b in let _ = putc b ']' in
  b.comma := true; b

let key b prefix =
  b.cnt := !(b.cnt) + 1; Printf.sprintf "%s_%d" prefix !(b.cnt)

let begin_ title =
  let b = { buf = Buffer.create 4096; depth = ref 0; comma = ref false; cnt = ref 0 } in
  let _ = puts b "{\"ui\":{" in
  b.depth := 2; b.comma := false;
  let _ = kv b "title" title in arr b "layout"

let end_layout b =
  let _ = close_arr b in b.comma := false; let _ = putc b ',' in
  let _ = puts b "\"translations\":{}," in
  let _ = puts b "\"userAccess\":{\"cancel\":true}" in
  let _ = puts b "}}" in Buffer.contents b.buf

let fieldset b title =
  let _ = obj b in
  let _ = kv b "type" "FIELDSET" in let _ = kv b "key" (key b "fs") in
  let _ = kv b "title" title in arr b "content"

let end_fieldset b = close_obj (close_arr b)

let label b text =
  let _ = obj b in
  let _ = kv b "type" "LABEL" in let _ = kv b "key" (key b "l") in
  close_obj (kv b "label" text)

let alert b msg color =
  let _ = obj b in
  let _ = kv b "type" "ALERT" in let _ = kv b "key" (key b "a") in
  close_obj (kv b "color" color (kv b "message" msg))

let badge b title color =
  let _ = obj b in
  let _ = kv b "type" "BADGE" in let _ = kv b "key" (key b "bd") in
  close_obj (kv b "color" color (kv b "title" title))

let spacer b px =
  let _ = obj b in
  let _ = kv b "type" "SPACER" in let _ = kv b "key" (key b "sp") in
  close_obj (kvint b "width" px)

let input b id label required =
  let _ = obj b in
  let _ = kv b "type" "INPUT" in let _ = kv b "key" (key b "i") in
  let b = kv b "label" label (kv b "id" id) in
  close_obj (if required then kvbool b "required" else b)

let checkbox b id label =
  let _ = obj b in
  let _ = kv b "type" "CHECKBOX" in let _ = kv b "key" (key b "cb") in
  close_obj (kv b "label" label (kv b "id" id))

let textarea b id label rows =
  let _ = obj b in
  let _ = kv b "type" "TEXTAREA" in let _ = kv b "key" (key b "ta") in
  let b = kv b "label" label (kv b "id" id) in
  close_obj (if rows > 0 then kvint b "rows" rows else b)

let rating b id label =
  let _ = obj b in
  let _ = kv b "type" "RATING" in let _ = kv b "key" (key b "rt") in
  close_obj (kv b "label" label (kv b "id" id))

let button b id title color is_default =
  let _ = obj b in
  let _ = kv b "type" "BUTTON" in let _ = kv b "key" (key b "btn") in
  let b = kv b "color" color (kv b "title" title (kv b "id" id)) in
  close_obj (if is_default then kvbool b "default" else b)

let section b title text = end_fieldset (label (fieldset b title) text)

File: DynamicLayout.scala
// DynamicLayout Scala API — generate layout JSON from Scala code.
// Single-file. Zero dependencies outside stdlib. Works on Scala 2.13+ / 3.x.

package dynamiclayout

class Builder(private val sb: StringBuilder = new StringBuilder(), private var depth: Int = 0, private var comma: Boolean = false, private var cnt: Int = 0):
  private def putc(c: Char) = { sb.append(c); this }
  private def puts(s: String) = { sb.append(s); this }
  private def indent = { putc('\n'); (0 until depth).foreach(_ => puts("  ")); this }
  private def cm = { if comma then putc(',') else this; comma = false; this }
  private def esc(s: String) = { putc('"'); s.foreach { case '"' => puts("\\\""); case '\\' => puts("\\\\"); case '\n' => puts("\\n"); case c => putc(c) }; putc('"') }
  private def kv(k: String, v: String) = { cm; putc(','); indent; esc(k); puts(": "); esc(v); comma = true; this }
  private def kvint(k: String, v: Int) = { cm; putc(','); indent; esc(k); puts(s": $v"); comma = true; this }
  private def kvbool(k: String) = { cm; putc(','); indent; esc(k); puts(": true"); comma = true; this }
  private def obj = { cm; putc(','); indent; putc('{'); depth += 1; comma = false; this }
  private def closeObj = { depth -= 1; indent; putc('}'); comma = true; this }
  private def arr(k: String) = { cm; putc(','); indent; esc(k); puts(": ["); depth += 1; comma = false; this }
  private def closeArr = { depth -= 1; indent; putc(']'); comma = true; this }
  private def key(p: String) = { cnt += 1; s"${p}_$cnt" }

  def fieldset(title: String): Builder = { obj; kv("type", "FIELDSET"); kv("key", key("fs")); kv("title", title); arr("content") }
  def endFieldset: Builder = { closeArr; closeObj }
  def row: Builder = { obj; kv("type", "ROW"); kv("key", key("r")); arr("content") }
  def endRow: Builder = { closeArr; closeObj }
  def label(text: String): Builder = { obj; kv("type", "LABEL"); kv("key", key("l")); kv("label", text); closeObj }
  def alert(msg: String, color: String = "info"): Builder = { obj; kv("type", "ALERT"); kv("key", key("a")); kv("message", msg); kv("color", color); closeObj }
  def badge(title: String, color: String = "primary"): Builder = { obj; kv("type", "BADGE"); kv("key", key("bd")); kv("title", title); kv("color", color); closeObj }
  def spacer(px: Int = 20): Builder = { obj; kv("type", "SPACER"); kv("key", key("sp")); kvint("width", px); closeObj }
  def input(id: String, label: String, required: Boolean = false): Builder =
    val b = { obj; kv("type", "INPUT"); kv("key", key("i")); kv("id", id); kv("label", label) }
    if required then b.kvbool("required").closeObj else b.closeObj
  def checkbox(id: String, label: String = ""): Builder = { obj; kv("type", "CHECKBOX"); kv("key", key("cb")); kv("id", id); kv("label", label); closeObj }
  def textarea(id: String, label: String = "", rows: Int = 3): Builder =
    val b = { obj; kv("type", "TEXTAREA"); kv("key", key("ta")); kv("id", id); kv("label", label) }
    if rows > 0 then b.kvint("rows", rows).closeObj else b.closeObj
  def rating(id: String, label: String = ""): Builder = { obj; kv("type", "RATING"); kv("key", key("rt")); kv("id", id); kv("label", label); closeObj }
  def button(id: String, title: String, color: String = "primary", isDefault: Boolean = false): Builder =
    val b = { obj; kv("type", "BUTTON"); kv("key", key("btn")); kv("id", id); kv("title", title); kv("color", color) }
    if isDefault then b.kvbool("default").closeObj else b.closeObj
  def section(title: String, text: String): Builder = fieldset(title).label(text).endFieldset
  def result: String = sb.toString()

object Layout:
  def begin(title: String): Builder =
    val b = Builder()
    b.puts("{\"ui\":{"); b.depth = 2; b.comma = false
    b.kv("title", title); b.arr("layout")
  def end(b: Builder): String =
    b.closeArr; b.comma = false; b.putc(',')
    b.puts("\"translations\":{},")
    b.puts("\"userAccess\":{\"cancel\":true}")
    b.puts("}}"); b.result
